### DISCUSSION TITLE
Using ui.download to download bytes?

### DISCUSSION THREAD

**Discussion by danpf** [zauberzeug/nicegui#828] [Q&A] [answered]
### Question

What would be the 'best practices' way to use ui.download to download a binary?  I feel like this is something that should be built into ui.download, but as that's very new, I was wondering what I should be using in the meantime?

I wrote a little example here:

```py
import tempfile
from pathlib import Path

from nicegui import app, ui


def ui_download_binary(local_dir: Path, data: bytes, filename: str):
    app.add_static_files(str(local_dir), str(local_dir))
    out_binary = local_dir / filename
    with open(out_binary, "wb") as fh:
        fh.write(data)
    ui.download(str(out_binary))


def run_computation(col: ui.column, temp_dir: str):
    # Stuff that takes a long time
    data_out = b"DATA11011"
    col.clear()
    with col:  # Bug here? in context manager, download happens multiple times...
        ui.button("Download Results", on_click=lambda: ui_download_binary(Path(temp_dir), data_out, "data.csv"))


def test_download():
    temp_dir = tempfile.mkdtemp() + "/"
    ui.button("Run", on_click=lambda: run_computation(col, temp_dir))
    col = ui.column()


if __name__ in {"__main__", "__mp_main__"}:
    test_download()
    ui.run()
```

**Comment by rodja** [comment, marked as answer]
Currently ui.download only accepts a url to download things. So you need to make your computed data available via [ui.add_static_files](https://nicegui.io/documentation#add_static_files_).

**Reply by danpf** [reply]
that's what I did here. Do you know why, when I use the col context manager (`with col`) and then click on the download button it downloads the file 3-4 times?

**Comment by rodja** [comment]
I just created a [feature request to allow passing of local paths](https://github.com/zauberzeug/nicegui/discussions/833). Maybe you or someone else can implement it?

**Reply by rodja** [reply]
Oh, it took me a while to understand that you edited the original posted code. Please do not do that. Editing is only to fix typos. If you change the original meaning, all discussions below become irrelevant / inaccurate. It's much better to post new messages.

**Reply by rodja** [reply]
About your problem with multiple downloads: I could not reproduce the issue. Could you create a new Q/A discussion or issue with a minimal reproducible  example?
